### ISSUE TITLE
[scan] `Route.graph()` is always a path, so engin-pathway's message passing is a six-entry lookup on `n_steps` — replacing Â with the identity raises Spearman 0.866 → 0.882 (p=1e-16), and all 67 tests pass with the hidden block zeroed

### ISSUE THREAD

**Issue by dcbaugher** [enginbio/engin-suite#399]
**Lens:** Implementation efficiency (rotation idx 2) · **Scanned:** 2026-09-12

## What

`engin_pathway.schema.Route.graph()` (`schema.py:220`) returns `nx.path_graph(self.n_steps)`
unconditionally. No other code path in the suite adds an edge to a route — `grep` for
`add_edge|add_edges_from` across all six packages returns exactly one site, and it is
`engin_materials/schema.py:97`. So `normalized_adjacency(g)` (`engin_graph/featurize.py:68`)
is a function of `n_steps` alone. Measured over `make_dataset(500)`:

```
distinct Â matrices across 500 routes: 6 (one per L in [2,3,4,5,6,7]), deviations: 0/500
```

Â is a six-entry lookup table keyed on the step count — which is exactly the signal the
step-count baseline already has. The GCN block `H2 = relu(Â relu(Â X W1) W2)`
(`engin_graph/embed.py:64-66`) therefore carries no graph information at all on this domain.

**The operator is not merely inert here. It costs accuracy.** Keep the same random `W1`/`W2`
and the same ReLUs; replace Â with the identity, deleting only the message passing. 40 data
seeds × 5 embed seeds, 500 routes on the README's own 320/90/90 split:

| variant | Spearman | R² (mean / sd / min) |
|---|---|---|
| shipped, `H2 = relu(Â relu(Â X W1) W2)` | 0.8659 ± 0.0328 | 0.627 / 1.045 / **−12.54** |
| **Â → I, same weights and ReLUs** | **0.8819 ± 0.0302** | **0.736** / 0.283 / −3.15 |

Paired difference −0.0159; the shipped operator loses on **148 of 200** runs; Wilcoxon
signed-rank **p = 1.4e-16**.

What the hidden block contributes is a random nonlinear feature expansion — Rahimi & Recht
random features, essentially — and the graph operator subtracts from it. Note the second
column: the shipped variant's R² has sd 1.045 and reaches **−12.54** on some (data seed,
embed seed) pairs, while Spearman on those same runs stays above 0.85, so a ranking-only
check never sees it. The README publishes **R² 0.73** (`README.md:27`), which is what the
*identity* variant delivers on average and not what ships.

**The closing argument is the test suite.** Zero the hidden block entirely
(`H2 = np.zeros_like(...)`) and `packages/engin-pathway` runs **67 passed, 4 skipped** — byte
for byte the result of the unmodified tree, `test_graph_model_outranks_step_count` and
`test_best_route_selection_beats_step_count` included.

## Why it matters

engin-pathway is funnel stage **[3]**, and the README states the mechanism as fact in the
present tense, under "What it does":

> embeds it with a **message-passing GCN** whose max/min-pooling captures the *worst step*
> — the structural signal step-count misses (`README.md:19`)

That attribution is wrong today, and the repository's own tests say so. The pooling that
captures the worst step is over the **raw** features, before any message passing:
`test_embed.py:33` asserts `toxic[10:15].min() < healthy[10:15].min()`, and with `d_in=5`
slots `10:15` are precisely the `X.min(0)` block (`embed.py:72`).
`engin_graph/tests/test_embed.py:43` makes the same assertion through `raw_min_block()`.
Both tests for "sees the worst step/node" index the pre-message-passing pool, because that
is where the signal is.

The cost is the ordinary D9 tally rather than a wrong number in the docs. `engin-graph`
carries `networkx>=3.0` and a dense `nx.to_numpy_array` per embed to build a matrix with six
possible values; `GCNEmbedder` carries two random weight matrices and a `seed`; and
`embed_seed` is a public constructor argument on `PathwayRanker` and `PolymerRanker`, so it
sits inside the `docs/api-stability.md` guarantee. On this domain all of that buys a
negative number.

## Evidence

- `packages/engin-pathway/src/engin_pathway/schema.py:220` — `nx.path_graph(self.n_steps)`,
  unconditional. `packages/engin-materials/src/engin_materials/schema.py:96-97` — the only
  place in the suite that adds an edge.
- `packages/engin-graph/src/engin_graph/embed.py:64-66` (Â, H1, H2), `:72` (`X.min(0)`),
  `:50-51` (random `W1`, `W2`); `packages/engin-graph/src/engin_graph/featurize.py:68`
  (`normalized_adjacency`).
- `packages/engin-pathway/tests/test_embed.py:33` and
  `packages/engin-graph/tests/test_embed.py:43` — the worst-step/worst-node assertions,
  both against the raw min-pool.
- `packages/engin-pathway/README.md:19` (the mechanism claim), `:27` (R² 0.73), `:28`
  (Spearman 0.85 vs 0.51).
- Rahimi & Recht, *Weighted Sums of Random Kitchen Sinks*, NIPS 2008 —
  [PDF](https://people.eecs.berkeley.edu/~brecht/papers/08.rah.rec.nips.pdf). The theory
  under "random projection + trained linear readout works": risk gap bounded as
  `O(1/√m + 1/√K)`. It is the honest description of what the hidden block does here, and it
  says nothing about graphs.
- Gallicchio & Micheli, *Fast and Deep Graph Neural Networks*, AAAI 2020 34(04):3898–3905,
  [10.1609/aaai.v34i04.5803](https://doi.org/10.1609/aaai.v34i04.5803),
  [arXiv:1911.08941](https://arxiv.org/abs/1911.08941) — the refereed prior art for untrained
  message passing with a trained readout. Named here because the README's "random-weight GCN"
  currently cites nothing, and because FDGNN's result is obtained on graphs whose topology
  varies, which is the condition this domain does not meet.
- Morris, Kriege, Bause, Kersting, Mutzel & Neumann, *TUDataset*, GRL+ @ ICML 2020,
  [arXiv:2007.08663](https://arxiv.org/abs/2007.08663) — "classical graph kernels in
  combination with SVMs are still highly competitive in graph classification tasks",
  best-overall on small-scale datasets. Adjacent rather than on-point: it is about trained
  GNNs on varied topologies, not about this construction.

## What it would take

The smallest version that settles it is the identity ablation above, as a benchmark under
`packages/engin-pathway/benchmarks/`, plus a present-tense correction to `README.md:19`
saying the worst-step signal comes from the raw min-pool and that route topology is currently
a constant function of length. That is a half-day and it does not touch the engine.

The larger question is which of two things is wrong, and it should be argued separately:
either `Route` should stop modelling every route as a chain — real routes branch, share
cofactors, and converge, and `schema.py`'s own M1 plan (Rhea ingest) would supply that — or
`engin-pathway` should not be paying for a graph engine until it has a graph. The first is
the interesting one. Do not delete `GCNEmbedder`: on `engin-materials`, where `crosslinks`
make the topology genuinely vary, it beats raw pooling by +0.084 across 30/30 seeds — see the
sibling issue filed today, which finds a different problem there.

## Decisions touched

`D9` — implements, in the awkward direction. `D9` says build only what has no open
equivalent; here the open equivalent is `numpy.min`. `D14` — adjacent: `embed_seed` is public
API surface on two domain classes for a block that contributes nothing on one of them.
ADR 0002 (light default dependency path) — neutral; nothing here argues for taking torch, and
the finding is that the torch-free stand-in is not doing the job the docs credit it with.
`DECISIONS.md` contains no entry on the graph embedder, and `AGENTS.md`'s "looks like a bug,
isn't" list does not reach it, so nothing is being routed around.

## Why this might be wrong

**My first version of this finding was wrong, and an adversarial re-run overturned it.** I
originally measured the whole hidden block against raw pooling plus `n_steps` at 8 data seeds
and got a paired difference of −0.0031, negative on 5 of 8 — "the block buys nothing". At 120
data seeds it reverses: the block is worth **+0.0053**, winning 72/120, Wilcoxon p=0.0012,
and the effect *grows* with data (n=1000: +0.0091, p<0.0001). The eight-seed result was an
artifact. The block is not useless; **the operator inside it is.** I am recording this
because the same underpowering would have produced a confident wrong issue, and because the
number above (−0.0159 at 200 runs, p=1.4e-16) is the one that survived the same treatment.

I also originally led with a 3× best-of-6 regret advantage for the shipped embedding
(0.0013 vs 0.0042). **That is a degenerate metric on this test set** — `best_of_k_regret` is
exactly 0.0 in roughly 97% of runs and the two variants differ in about a dozen runs out of
400. Under the demo's actual protocol it is p=0.53. It should not have been tabled and is not
tabled here.

**The strongest surviving objection is the M1 defence.** `engin-graph/README.md` already says
"Do not read M0 ranking numbers as a claim about real-world accuracy", the pathway README
calls the random weights "a deliberate stand-in" (`:84`), and M1 swaps in a trained GNN on
real routes, which *are* branched. On that reading: of course message passing is inert on a
chain generator, and this measures the stand-in rather than the design. I think that blunts
the numbers and does not touch the two facts that matter — the README describes the mechanism
in the present tense under "What it does", and the operator is not neutral but negative, which
is a property of the operator plus this pooling and not of training. It is a reason to check
the M1 design, not to wave the current sentence through.

The second objection is that `nx.path_graph` is deliberate and obvious, so none of this is
news. It is obvious from `schema.py:220`; what is not obvious, and what I could not find
recorded anywhere, is that the consequence is measurable and negative. #124 audits the
generator's constants and quotes the adjacent README sentence while doing so; #88 and the
materials README audit topology-vs-composition; `sources.yaml`'s "Route-as-graph embedding"
row covers PyG-as-standard-impl and ADR 0002. Two repository issue searches for an ablation of
this kind returned nothing.

The third is that Spearman is not the metric that matters and R² is unstable on both variants.
True — but the shipped variant's R² sd is 1.045 against 0.283, and the README publishes a
point value from it.
